core/store/src/lib.rs

```diff
@@ -291,28 +291,70 @@ impl NodeStorage {
     }
 }
 
+pub fn retrieve_shard_uid_from_db_key(key: &[u8]) -> io::Result<ShardUId> {
+    // TODO(reshardingV3) Consider changing the Error type to `StorageError`?
+    // Would need changing error types for `Store` methods as well.
+    ShardUId::try_from(&key[..8])
+        .map_err(|e| io::Error::other(format!("Could not retrieve ShardUId from db key: {}", e)))
+}
+
+pub fn replace_shard_uid_key_prefix(key: &[u8], shard_uid: ShardUId) -> [u8; 40] {
+    let mut mapped_key = [0u8; 40];
+    mapped_key[..8].copy_from_slice(&shard_uid.to_bytes());
+    mapped_key[8..].copy_from_slice(&key[8..]);
+    mapped_key
+}
+
 impl Store {
     pub fn new(storage: Arc<dyn Database>) -> Self {
         Self { storage }
     }
 
+    /// Underlying `get()` implementation for all columns.
+    fn get_impl(&self, column: DBCol, key: &[u8]) -> io::Result<Option<DBSlice<'_>>> {
+        let value = if column.is_rc() {
+            self.storage.get_with_rc_stripped(column, &key)
+        } else {
+            self.storage.get_raw_bytes(column, &key)
+        }?;
+        Ok(value)
+    }
+
+    /// Specialized `get` implementation for State column that replaces shard_uid prefix
+    /// with a mapped value according to mapping strategy in Resharding V3.
+    ///
+    /// It does extra read from `DBCol::ShardUIdMapping` to map the shard_uid key prefix.
+    fn get_impl_state(&self, key: &[u8]) -> io::Result<Option<DBSlice<'_>>> {
+        let shard_uid = retrieve_shard_uid_from_db_key(key)?;
+        let mapped_shard_uid = self
+            .get_ser::<ShardUId>(DBCol::ShardUIdMapping, &shard_uid.to_bytes())?
```

Comment:
Could we do the mapping in core/store/src/adapter/trie_store.rs ?

Reply:
Having `TrieStoreAdapter` is now so nice :)